ethereum/listener.go

```diff
@@ -10,187 +10,360 @@ import (
+//
+// If an error occurs in websocket stream, this function will handle relevant sub routines and then re-run itself.
+func (e *Ethereum) startListenerRoutines(
+	ctx context.Context,
+	logger log.Logger,
+) {
+
+	sig := &errSignal{
+		Ready: make(chan struct{}),
 	}
 
-	messageTransmitterAddress := common.HexToAddress(e.messageTransmitterAddress)
-	etherReader := etherstream.Reader{Backend: ethClient}
+	// start main stream (does not account for lookback period or specific start block)
+	stream, sub, history := e.startMainStream(ctx, logger)
 
-	if e.startBlock == 0 {
-		header, err := ethClient.HeaderByNumber(ctx, nil)
-		if err != nil {
-			logger.Error("unable to retrieve latest eth block header", "err", err)
-			os.Exit(1)
-		}
+	go e.consumeStream(ctx, logger, stream, sig)
+	consumeHistory(logger, history)
+
+	// get history from (start block - lookback) up until latest block
+	latestBlock := e.LatestBlock()
+	start := latestBlock
+	if e.startBlock != 0 {
+		start = e.startBlock
+	}
+	startLookback := start - e.lookbackPeriod
+	logger.Info(fmt.Sprintf("Getting history from %d: starting at: %d looking back %d blocks", startLookback, start, e.lookbackPeriod))
+	e.getAndConsumeHistory(ctx, logger, startLookback, latestBlock)
+
+	logger.Info("Finished getting history")
 
-		e.startBlock = header.Number.Uint64()
+	if flushInterval > 0 {
+		go e.flushMechanism(ctx, logger, sig)
 	}
 
+	// listen for errors in the main websocket stream
+	// if error occurs, trigger sig.Ready
+	// This will cancel `consumeStream` and `flushMechanism` routines
+	select {
+	case <-ctx.Done():
+		return
+	case err := <-sub.Err():
+		logger.Error("Websocket disconnected. Reconnecting...", "err", err)
+		close(sig.Ready)
+
+		// restart
+		e.startBlock = e.lastFlushedBlock
+		time.Sleep(10 * time.Millisecond)
+		e.startListenerRoutines(ctx, logger)
+		return
+	}
+
+}
+
+func (e *Ethereum) startMainStream(
+	ctx context.Context,
+	logger log.Logger,
+) (stream <-chan ethtypes.Log, sub ethereum.Subscription, history []ethtypes.Log) {
+
+	var err error
+
+	etherReader := etherstream.Reader{Backend: e.wsClient}
+
+	latestBlock := e.LatestBlock()
+
+	// start initial stream (start-block and lookback period handled separately)
+	logger.Info("Starting Ethereum listener")
+
 	query := ethereum.FilterQuery{
 		Addresses: []common.Address{messageTransmitterAddress},
 		Topics:    [][]common.Hash{{messageSent.ID}},
-		FromBlock: big.NewInt(int64(e.startBlock - e.lookbackPeriod)),
+		FromBlock: big.NewInt(int64(latestBlock)),
 	}
 
-	logger.Info(fmt.Sprintf(
-		"Starting Ethereum listener at block %d looking back %d blocks",
-		e.startBlock,
-		e.lookbackPeriod))
+	queryAttempt := 1
+	for {
+		// websockets do not query history
+		// https://github.com/ethereum/go-ethereum/issues/15063
+		stream, sub, history, err = etherReader.QueryWithHistory(ctx, &query)
+		if err != nil {
+			logger.Error("Unable to subscribe to logs", "attempt", queryAttempt, "err", err)
+			queryAttempt++
```

Comment:
if queryAttempt gets above a certain number should this ever panic? or just a descriptor for logging?

Reply:
My thinking here is that if for some reason the websocket is offline and it was unable to subscribe to logs, we should continue relaying and flushing other paths. 

This would be a good place to add prometheus metrics so that an infra team can be notified about these errors.